### ISSUE TITLE
[Proposal] Put each nimble package in its own file

### ISSUE THREAD

**Issue by jyapayne** [nim-lang/packages#777]
**Proposal**

Put each nimble package in its own file/directory, or let each user create a subdirectory for themselves and put each package into one file.

**Benefits**

- User only has to add a file to the repo instead of editing one big file. Multiple users can add packages without conflicts
- Parser won't eventually be overwhelmed by packages (since Nim and nimble will be super popular one day!)
- Better git history
- If each user has their own directory, packages can be found easily by the same user (not that it isn't easy already with nimble search, but just browsing the github repo could be a fun discovery)

**Downsides**

- Lots of files. But git can handle it.
- Higher code complexity. Must write recursive directory walker and gather packages.

So what do you think? It's just an idea, but maybe it'll be of use?

**Comment by Araq** [comment]
You are talking about ``packages.json``, right?

**Comment by jyapayne** [comment]
@Araq yep!

**Comment by dom96** [comment]
Sure, we can do this :)

**Comment by FedericoCeratto** [comment]
> Parser won't eventually be overwhelmed by packages
This is not correct. The most common searches are by package name, tag and words from the description.
Indexing the packages require scanning the whole packages.json file. With separate files this would require recursive walks and parsing a large number of manifest files which is even slower.

The usual solution for large archieves is to precompute index files centrally (essentially a keyword -> package name k/v map) and ship those. Yet, we are not going to need this for quite a while!

**Comment by jyapayne** [comment]
@FedericoCeratto You make a good point about the package scanner and it taking more time to do recursive walks, however, the main point (which I failed to emphasize) is the greater ease of use.

It's not happening so much now, but when lots of users start submitting lots of packages to Nim, there are potentially going to be many PRs for `packages.json` and many conflicts to resolve either by the user or by @dom96 and yourself, or whoever is maintaining it in the future.

Again, this is just an idea. If it makes things too difficult or my reasoning is faulty, maybe package.json fine the way it is. I haven't looked at the source, so maybe this is the most efficient way.

**Comment by dom96** [comment]
One potential drawback with this proposal is that you won't be able to as easily add a new package using just the GitHub web interface.

**Comment by jyapayne** [comment]
@dom96 that's true as well. It's pretty easy to just add a package like that.

**Comment by euantorano** [comment]
You can still fairly easily add new packages using the GitHub UI by clicking `Create new file`, you just have to then copy the JSON (or whatever) structure into the editor.

**Comment by jyapayne** [comment]
@euantorano that's a good point as well. I'm still a proponent of this change and I'll look into the effort when I have some time.

**Comment by FedericoCeratto** [comment]
@jyapayne I'll create a proposal on the Nimble issue tracker for a package publishing service.

**Comment by jyapayne** [comment]
@FedericoCeratto that would be equally awesome!

**Comment by yglukhov** [comment]
> One potential drawback with this proposal is that you won't be able to as easily add a new package using just the GitHub web interface.

Actually you can create files through web interface.

**Comment by genotrance** [comment]
You can also create a folder structure using tab though its not intuitive.

**Comment by FedericoCeratto** [comment]
**Problems**
The packages.json file is getting bigger. It could become slow to parse. It is awkward to edit. It requires contributors to have a GitHub account. It requires PR approval to prevent package hijacking.
Also: sometimes GH repos ar deleted, and some countries block GitHub, and GitHub refuses to serve some other countries.

**Proposal**
Precompute index files centrally: a <keyword> -> <package name> k/v map, to look up packages by keyword/tag; Also a <package name> -> <package metadata> map. Ship the two indexes as binary/compressed files for fast transfer and fast lookup time.

Run a simple service similar to pypi.org to handle package creation/update and generate the indexes. Initially it could feed from GH and/or use GH as a backend to store the indexes.

**future goals**
Store compressed tarballs of released packages. This is useful in case of dead repos, and in countries that block GitHub, and countries that GitHub refuses to serve.
Check URL / git repo existence before accepting a new package.

**moonshot goals**
Let package owners sign metadata. Also use the signature to allow allow owners to update/delete packages without having to store logins and passwords.
Verify signed tarballs from GH (and other sources) against the owner pubkey.
A pool of "admin" pubkeys is allowed to update/delete other packages.
A pool of "contributor" pubkeys can vet trusted packages by adding a "vote +1" signature. Nimble can warn before installing unvetted packages.

**Comment by dom96** [comment]
> A small database is preferred over a directory because it can be: downloaded easily over plain http, replaced atomically on disk, checksummed to verify integrity, signed.
> SQLite is supported by the stdlib, has a stable format, works cross-platform, has really good lookup timing.

Nothing is simpler than a directory structure. SQLite is a dependency and I would not wish to make Nimble depend on it.

Everything else sounds good to me on the surface. It'll be the details of how you implement this that I may disagree with, but just go for it :)

**Comment by ZeroAurora** [comment]
Any progress here? I'm learning Nim but found the package list is pretty ugly. Maybe we can separate it to single files using a script and compile them into a single json file using CI for backward compability. Use git or tarball to fetch the package repo is preferred by me though.
